File: raber.py
#!/usr/bin/env python
"""
raber.py

Given the number of sections needed, calculate the number 12m rebars
required and the amount of waste.
"""

import math
from collections import Counter, defaultdict
from itertools import combinations

REBAR_LENGTH = 1200
PIECE_SIZE = 550

sizes = {
    3: {
        170: 16,  # T02
        15: 8,  # T03
        200: 126,  # Decking!
    },
    4: {325: 20},  # A03
    5: {
        413: 8,  # T01
        300: 4,  # A02
    },
    6: {285: 20},  # A01
}


def calculate_waste(required: dict):
    """
    Calculate waste from rebar pieces.
    """
    total_pieces = sum(required.values())

    # Always assume that the full rebar produces two pieces
    num_full_rebars = math.ceil(total_pieces / 2)
    from_bend = (REBAR_LENGTH - (PIECE_SIZE*2)) * num_full_rebars

    from_extra_piece = PIECE_SIZE * (total_pieces % 2)

    from_piece_remainders = sum(
        (PIECE_SIZE - sum(k)) * v
        for k, v in required.items()
    )

    total_waste = from_bend + from_extra_piece + from_piece_remainders

    return num_full_rebars, total_waste


def create_combinations(needed: dict):
    """
    Assumption: There are only two kinds
    """
    def get_max_bars(tup: tuple):
        length, num_pcs = tup
        return [length] * min(math.ceil(PIECE_SIZE/length), num_pcs)
    all_valid_bars = [j for i in needed.items() for j in get_max_bars(i)]
    valid_combos = set()
    seen = set()
    for i in range(len(all_valid_bars)):
        for c in combinations(all_valid_bars, i):
            if c not in seen and sum(c) <= PIECE_SIZE and len(c):
                valid_combos.add(c)
            seen.add(c)
    return valid_combos


def combo_fits(needed: dict, combo: tuple):
    """
    Check if the combination satisfies the pieces still needed
    """
    c = Counter(combo)
    valid = all(c[k] <= needed[k] for k in c)
    return c if valid else None


def main(needed: dict):
    """
    Inputs
    ------

    - `input`: Dictionary with keys as lengths and values as the quantity.
    """

    # Case 1: the PIECE_SIZE can only produce 1 piece from the input,
    # the rest is wasted
    if min(needed.keys()) * 2 > PIECE_SIZE:
        return {(length,): num for length, num in needed.items()}

    # Case 2: Only one length, just calculate
    if len(needed) == 1:
        length, num_pcs = next(iter(needed.items()))
        num_in_1_piece = PIECE_SIZE // length
        num_rebars = {(length,) * num_in_1_piece: num_pcs//num_in_1_piece}
        remainder = num_pcs % num_in_1_piece
        if remainder:
            num_rebars[(length,) * remainder] = 1
        return num_rebars

    # Case 3: Can be matched
    # greedy solution, not sure if it's good tbh
    combinations = create_combinations(needed)
    cp = dict(needed)
    rebars = defaultdict(lambda: 0)
    while any(c for c in cp.values()):
        min_waste = max(combinations, key=sum)
        while (counter := combo_fits(cp, min_waste)):
            rebars[min_waste] += 1
            for c in counter:
                cp[c] -= counter[c]
        combinations.remove(min_waste)

    return dict(rebars)


if __name__ == "__main__":
    for k, v in sizes.items():
        num_rebars = main(v)
        num_full_rebars, waste = calculate_waste(num_rebars)
        print("-------------------\n"
              f"{v}\n"
              f"#{k} rebar {PIECE_SIZE} cm pieces: {num_rebars}\n"
              f"#{k} full 12m rebars: {num_full_rebars}\n"
              f"waste: {waste/100:02} m")

File: Raber.js
class Raber {
  #bendWaste;
  #rebarPieceSize;

  constructor(rebarLength, rebarPieceSize) {
    this.#bendWaste = rebarLength - (rebarPieceSize * 2);
    this.#rebarPieceSize = rebarPieceSize;
  }

  calculateWaste(required) {
    const req = Object.entries(required);
    const totalPieces = req
      .map((e) => e[1])
      .reduce((acc, curr) => acc + numOrNumList(curr), 0);

    // Always assume that the full rebar produces two pieces
    const numFullRebars = Math.ceil(totalPieces / 2);

    const fromBend = this.#bendWaste * numFullRebars;
    const fromExtraPiece = this.#rebarPieceSize * (totalPieces % 2);
    const fromPieceRemainders = req
      .reduce(
        (acc, [length, quantity]) =>
          acc + (this.#rebarPieceSize - numOrNumList(length)) * quantity,
        0,
      );

    return [numFullRebars, fromBend + fromExtraPiece + fromPieceRemainders];
  }

  createCombinations(entries) {
    const allValidBars = entries.flatMap(([length, numPieces]) =>
      makeArr(
        Math.min(Math.ceil(this.#rebarPieceSize / length), numPieces),
        length,
      )
    );

    const validCombos = new Set();
    const seen = new Set();

    for (let i = 0; i <= allValidBars.length; i++) {
      for (const combo of getPermutations(allValidBars, i)) {
        if (combo.length > 0 && !seen.has(JSON.stringify(combo))) {
          const sumLengths = combo.reduce((acc, curr) => acc + curr, 0);
          if (sumLengths <= this.#rebarPieceSize) {
            validCombos.add(JSON.stringify(combo.sort()));
          }
        }
        seen.add(JSON.stringify(combo));
      }
    }

    return Array.from(validCombos, (combo) => JSON.parse(combo));
  }

  calculate(needed) {
    const entries = Object.entries(needed).map(([k, v]) => [+k, +v]);

    // Case 1: Only 1 type of section piece
    if (entries.length === 1) {
      const [[length, numPieces]] = entries;
      return singleType(numPieces, length, this.#rebarPieceSize);
    }

    // Case 2: Only 1 section piece can be produced from the rebar piece
    if (Math.min(entries.map((e) => e[0])) * 2 > this.#rebarPieceSize) {
      return entries.reduce((acc, [length, num]) => {
        acc[[length]] = num;
        return acc;
      }, {});
    }

    let combinations = Array.from(this.createCombinations(entries));
    const cp = { ...needed };
    const rebars = {};

    while (Object.values(cp).some((v) => v > 0)) {
      const minWaste = combinations.reduce((acc, curr) =>
        sum(curr) > sum(acc) ? curr : acc
      );

      let counter;
      while ((counter = comboFits(cp, minWaste))) {
        rebars[JSON.stringify(minWaste)] =
          (rebars[JSON.stringify(minWaste)] || 0) + 1;
        for (const [length, count] of counter) {
          cp[length] -= count;
        }
      }
      combinations = combinations.filter((combo) => !isEqual(combo, minWaste));
    }
    return rebars;
  }
}

function isEqual(arr1, arr2) {
  return arr1.length === arr2.length &&
    arr1.every((val, index) => val === arr2[index]);
}

function sum(arr) {
  return arr.reduce((acc, curr) => acc + curr, 0);
}

function comboFits(needed, combo) {
  const c = combo.reduce((counter, length) => {
    counter[length] = (counter[length] || 0) + 1;
    return counter;
  }, {});

  const entries = Object.entries(c);
  const valid = entries.every(([length, count]) => count <= needed[length]);
  return valid ? entries : null;
}

function getPermutations(array, size) {
  function p(t, i) {
    if (t.length === size) {
      result.push(t);
      return;
    }
    if (i + 1 > array.length) {
      return;
    }
    p(t.concat(array[i]), i + 1);
    p(t, i + 1);
  }

  var result = [];
  p([], 0);
  return result;
}

function makeArr(length, item) {
  return Array.from({ length }, () => item);
}

function singleType(numPieces, pieceLength, rebarPieceSize) {
  const numInOnePiece = Math.floor(rebarPieceSize / pieceLength);
  const numRebars = {};
  const multiple = Math.floor(numPieces / numInOnePiece);
  if (multiple > 0) {
    numRebars[JSON.stringify(makeArr(numInOnePiece, pieceLength))] = multiple;
  }

  const remainder = numPieces % numInOnePiece;
  if (remainder > 0) {
    numRebars[JSON.stringify(makeArr(remainder, pieceLength))] = 1;
  }
  return numRebars;
}

function numOrNumList(v) {
  const value = JSON.parse(String(v));
  if (Array.isArray(value)) {
    return sum(value);
  } else {
    return value;
  }
}
